tools/block-generator/generator/generate_apps.go

```diff
@@ -0,0 +1,260 @@
+		// can only override the appID when non-zero in hintApp
+		appID = hintApp.appID
+	}
+	// WLOG: the matched cases below are now well-defined thanks to getActualAppCall()
+
+	var signedTxns []txn.SignedTxn
+	switch appCallType {
+	case appTxTypeCreate:
+		appID = g.txnCounter + intra + 1
+		signedTxns = g.makeAppCreateTxn(kind, senderAcct, round, intra, appID)
+		reSignTxns(signedTxns)
+
+		for k := range g.appMap {
+			if g.appMap[k][appID] != nil {
+				return "", nil, appID, fmt.Errorf("should never happen! app %d already exists for kind %s", appID, k)
+			}
+			if g.pendingAppMap[k][appID] != nil {
+				return "", nil, appID, fmt.Errorf("should never happen! app %d already pending for kind %s", appID, k)
+			}
+		}
+
+		ad := &appData{
+			appID:  appID,
+			sender: senderIndex,
+			kind:   kind,
+			optins: map[uint64]bool{},
+		}
+
+		g.pendingAppSlice[kind] = append(g.pendingAppSlice[kind], ad)
+		g.pendingAppMap[kind][appID] = ad
+
+	case appTxTypeOptin:
+		signedTxns = g.makeAppOptinTxn(senderAcct, round, intra, kind, appID)
+		reSignTxns(signedTxns)
+		if g.pendingAppMap[kind][appID] == nil {
+			ad := &appData{
+				appID:  appID,
+				sender: senderIndex,
+				kind:   kind,
+				optins: map[uint64]bool{},
+			}
+			g.pendingAppMap[kind][appID] = ad
+			g.pendingAppSlice[kind] = append(g.pendingAppSlice[kind], ad)
+		}
+		g.pendingAppMap[kind][appID].optins[senderIndex] = true
+
+	case appTxTypeCall:
+		signedTxns = []txn.SignedTxn{
+			signTxn(g.makeAppCallTxn(senderAcct, round, intra, appID)),
+		}
+
+	default:
+		return "", nil, appID, fmt.Errorf("unimplemented: invalid transaction type <%s> for app %d", appCallType, appID)
+	}
+
+	return actual, signedTxns, appID, nil
+}
+
+func (g *generator) getAppData(existing bool, kind appKind, senderIndex, appID uint64) (*appData, bool /* appInMap */, bool /* senderOptedin */) {
+	var appMapOrPendingAppMap map[appKind]map[uint64]*appData
+	if existing {
+		appMapOrPendingAppMap = g.appMap
+	} else {
+		appMapOrPendingAppMap = g.pendingAppMap
+	}
+
+	ad, ok := appMapOrPendingAppMap[kind][appID]
+	if !ok {
+		return nil, false, false
+	}
+	if !ad.optins[senderIndex] {
+		return ad, true, false
+	}
+	return ad, true, true
+}
+
+// getActualAppCall returns the actual transaction type, app kind, app transaction type and appID
+// * it returns actual = txType if there aren't any problems (for example create always is kept)
+// * it creates the app if the app of the given kind doesn't exist
+// * it switches to noopoc instead of optin when already opted into existing apps
+// * it switches to create instead of optin when only opted into pending apps
+// * it switches to optin when noopoc if not opted in and follows the logic of the optins above
+// * the appID is 0 for creates, and otherwise a random appID from the existing apps for the kind
+func (g *generator) getActualAppCall(txType TxTypeID, senderIndex uint64) (TxTypeID, appKind, appTxType, uint64 /* appID */, error) {
```

Comment:
## A note of explanation regarding this complicated function

Sometimes the generator receives a request to generate a particular transaction type which it cannot fulfill. This is true for asset and application transactions. For example one may receive an asset txn request when an asset doesn't yet exist or the sender isn't opted in. Similarly, one may receive an app call request when the app doesn't yet exist. Apps are more subtle as some apps need to be opted into while others not. So this is an inherently complicated situation. This complexity for assets is reflected in the method `generateAssetTxnInternalHint()` which uses recursion to "try again" when a particular txn type isn't yet possible. Similarly, `generateAppCallInternal()` also needs to switch an `actual` txn for the requested. However, as opposed to using recursion directly, it first figures out exactly what the `actual` should be, and then delegates to a `createXYZtransaction()` method. It figures out the `actual` using this method `getActualAppCall()`. This method uses recursion for the same reason that `generateAssetTxnInternalHint()` does.

Reply:
Nice comment.

The explanation is also a nice overview of this kind of "best effort" algorithm to meet the requested ratios.

Payments might even have similar logic in the event it needs to create a new account.

Reply:
I'm adding a TODO that the spirit of this comment needs to be added to the README, along with an in-depth explanation of how the block generator works. If we end up refactoring in short order to use the ledger, it might be worth delaying the README work though.